packages/store-sync/src/wiresaw.ts

```diff
@@ -0,0 +1,126 @@
+import { Observable } from "rxjs";
+import { Hex, LogTopic, RpcLog, encodeEventTopics, formatLog, parseEventLogs, toHex } from "viem";
+import { StorageAdapterBlock, StoreEventsLog } from "./common";
+import { storeEventsAbi } from "@latticexyz/store";
+import { logSort } from "@latticexyz/common";
+import { SocketRpcClient, getWebSocketRpcClient } from "viem/utils";
+
+type WatchLogsInput = {
+  url: string;
+  fromBlock: bigint;
+  address?: Hex;
+};
+
+type WatchLogsResult = {
+  logs$: Observable<StorageAdapterBlock>;
+};
+
+export function watchLogs({ url, address, fromBlock }: WatchLogsInput): WatchLogsResult {
+  // Buffer the live logs received until the gap from `startBlock` to `currentBlock` is closed
+  let caughtUp = false;
+  const logBuffer: StoreEventsLog[] = [];
+
+  const topics = [
+    storeEventsAbi.flatMap((event) => encodeEventTopics({ abi: [event], eventName: event.name })),
+  ] as LogTopic[]; // https://github.com/wevm/viem/blob/63a5ac86eb9a2962f7323b4cc76ef54f9f5ef7ed/src/actions/public/getLogs.ts#L171
+
+  const logs$ = new Observable<StorageAdapterBlock>((subscriber) => {
+    let client: SocketRpcClient<WebSocket>;
+    getWebSocketRpcClient(url, { keepAlive: true, reconnect: true }).then(async (_client) => {
+      client = _client;
+      client.socket.addEventListener("error", (error) =>
+        subscriber.error({ code: -32603, message: "WebSocket error", data: error }),
+      );
+
+      // Start watching pending logs
+      const subscriptionId: Hex = (
+        await client.requestAsync({
+          body: {
+            method: "wiresaw_watchLogs",
+            params: [{ address, topics }],
+          },
+        })
+      ).result;
+
+      // Listen for wiresaw_watchLogs subscription
+      // Need to use low level methods since viem's socekt client only handles `eth_subscription` messages.
+      // (https://github.com/wevm/viem/blob/f81d497f2afc11b9b81a79057d1f797694b69793/src/utils/rpc/socket.ts#L178)
+      client.socket.addEventListener("message", (message) => {
+        const response = JSON.parse(message.data);
+        if ("error" in response) {
```

Comment:
```suggestion
        if ("error" in response) {
          // if error isn't associated with a request ID, ignore
          // TODO: keep track of our request IDs and ignore others?
          if ("id" in response && response.id != null) return;
```

Reply:
actually could maybe lift this early return outside the error handler?

Reply:
I discovered that something sending other random RPC calls (`net_version` specifically, so I think it was Metamask) to the websocket will cause this to close

Reply:
it should currently open a separate websocket connection for this log stream, did you experience any other errors come through here?

Reply:
yes, I saw `net_version` occasionally sent through this (and not sure where from or how it got a reference to this open socket) and that would fail on our node, causing the stream to close, until it was recently added to the whitelist

Reply:
Turns out `net_version` request comes from viem's `ping` method used by `keepAlive` option. The right move was to enable this at our node rather than fixing anything here.

And it would be good if viem keeps `net_version` as the ping method (rather than e.g. moving to `eth_chainId`) because we cache `eth_chainId` requests to avoid it mucking with the transaction lifecycle and to reduce end-to-end latency.